src/rdyadvance.c

```diff
@@ -169,34 +165,143 @@ static PetscErrorCode CreateVizViewerContext(RDy rdy, ViewerContext *viz) {
       }
     }
   }
 
   PetscFunctionReturn(0);
 }
 
-PetscErrorCode RDyRun(RDy rdy) {
+// this is called when -preload is set so we can get more accurate timings
+// for the solver we're using
+static PetscErrorCode CalibrateSolverTimers(RDy rdy) {
+  PetscFunctionBegin;
+
+  // create a "preload" solution so we can advance one step
+  Vec X_preload;
+  PetscCall(VecDuplicate(rdy->X, &X_preload));
+  PetscCall(VecCopy(rdy->X, X_preload));
+
+  // set tolerances to make the calibration step cheaper
+  SNES      snes;
+  PetscReal r_tol;
+  PetscCall(TSGetSNES(rdy->ts, &snes));
+  PetscCall(SNESGetTolerances(snes, NULL, &r_tol, NULL, NULL, NULL));
+  PetscCall(SNESSetTolerances(snes, PETSC_DEFAULT, .99, PETSC_DEFAULT, PETSC_DEFAULT, PETSC_DEFAULT));
+
+  // take the step
+  PetscCall(TSSetSolution(rdy->ts, X_preload));
+  PetscCall(TSStep(rdy->ts));
+
+  // reset the tolerances and clean up
+  PetscCall(SNESSetTolerances(snes, PETSC_DEFAULT, r_tol, PETSC_DEFAULT, PETSC_DEFAULT, PETSC_DEFAULT));
+  PetscCall(VecDestroy(&X_preload));
+
+  PetscFunctionReturn(0);
+}
+
+/// Advances the solution by the coupling interval specified in the input
+/// configuration.
+PetscErrorCode RDyAdvance(RDy rdy) {
   PetscFunctionBegin;
 
-  PetscCall(CreateOutputDir(rdy));
+  // if we're at the start of the simulation, set up monitoring
+  if (rdy->step == 0) {
+    PetscCall(CreateOutputDir(rdy));
 
-  // set up monitoring functions for handling restarts and outputs
-  //  if (rdy->config.restart_frequency) {
-  //    PetscCall(TSMonitorSet(rdy->ts, WriteRestartFiles, rdy, NULL));
-  //  }
+    // set up monitoring functions for handling restarts and outputs
+    // if (rdy->config.restart_frequency) {
+    //   PetscCall(TSMonitorSet(rdy->ts, WriteRestartFiles, rdy, NULL));
+    // }
 
-  // create a viewer with the proper format for visualization output
-  ViewerContext viz = {0};
-  PetscCall(CreateVizViewerContext(rdy, &viz));
+    // create a viewer with the proper format for visualization output
+    PetscCall(CreateOutputViewer(rdy));
 
-  // do the thing!
-  RDyLogDebug(rdy, "Running simulation...");
-  PetscCall(TSSolve(rdy->ts, rdy->X));
+    RDyLogDebug(rdy, "Running simulation...");
+  }
+
+  // advance the solution to the specified time
+  PetscPreLoadBegin(PETSC_FALSE, "RDyAdvance solve");  // <-- enable with -preload true
+  PetscCall(TSSetTime(rdy->ts, rdy->t));
+  PetscReal interval = ConvertTimeToSeconds(rdy->config.time.coupling_interval, rdy->config.time.unit);
+  RDyLogDetail(rdy, "Advancing from t = %g to %g...", ConvertTimeFromSeconds(rdy->t, rdy->config.time.unit),
+               ConvertTimeFromSeconds(rdy->t + interval, rdy->config.time.unit));
+  PetscCall(TSSetMaxTime(rdy->ts, rdy->t + interval));
+  PetscCall(TSSetStepNumber(rdy->ts, rdy->step));
+  PetscCall(TSSetTimeStep(rdy->ts, rdy->dt));
+  PetscCall(TSSetSolution(rdy->ts, rdy->X));
+  if (PetscPreLoadingOn) {
```

Comment:
I believe 

```
if (PetscPreLoadingOn) {
```

should be changed to

```
if (PetscPreLoadingOn and rdy->step == 0) {
```

otherwise each time E3SM calls `RDyAdvance()`, we will end up reloading the initial condition.

Reply:
May be I have missed the whole point about preloading the vector.

Reply:
From the previous discussion, I got the impression that preloading is about getting more accurate timings by making sure the TS timers are initialized by the time we do the first solve.

Reply:
Preloading is trying to make sure that data is "hot". So we compute with quantities once and then immediately do it again. Assuming things are large, the caches are not important, but the data can be paged into our TLB.

Reply:
Preloading is generally not interesting for a standard production run. It's super valuable when you want to get reliable timing from profiling a small number of time steps. Preloading will only be on for one loop so I don't think you need or want to have the `rdy->step` conditional. Also, how do you envision doing restarts from a checkpoint? At least in TS, you will start at a nonzero step number.

(In general, I find mirroring things like `rdy->step` is bugprone compared to having a single source of authority.)